libraries/AC_PID/AC_PID.cpp

```diff
@@ -73,108 +87,149 @@ void AC_PID::set_dt(float dt)
+//  Target and Measured must be set manually for logging purposes.
+// todo: remove function when it is no longer used.
+float AC_PID::update_error(float error, bool limit)
 {
     // don't process inf or NaN
-    if (!isfinite(input)) {
-        return;
+    if (!isfinite(error)) {
+        return 0.0f;
     }
 
+    _target = 0.0f;
+
     // reset input filter to value received
     if (_flags._reset_filter) {
         _flags._reset_filter = false;
-        _input = input;
+        _error = error;
         _derivative = 0.0f;
+    } else {
+        float error_last = _error;
+        _error += get_filt_E_alpha() * (error - _error);
+
+        // calculate and filter derivative
+        if (_dt > 0.0f) {
+            float derivative = (_error - error_last) / _dt;
+            _derivative = _derivative + get_filt_D_alpha() * (derivative - _derivative);
+        }
     }
 
     // update filter and calculate derivative
-    if (_dt > 0.0f) {
-        float derivative = (input - _input) / _dt;
-        _derivative = _derivative + get_filt_alpha() * (derivative-_derivative);
-    }
+    update_i(limit);
 
-    _input = input;
-}
+    _pid_info.target = 0.0f;
+    _pid_info.actual = 0.0f;
+    _pid_info.error = _error;
+    _pid_info.P = (_error * _kp);
+    _pid_info.D = (_derivative * _kd);
 
-float AC_PID::get_p()
-{
-    _pid_info.P = (_input * _kp);
-    return _pid_info.P;
+    return (_error * _kp) + _integrator + (_derivative * _kd);
 }
 
-float AC_PID::get_i()
+//  update_i - update the integral
+//  If the limit flag is set the integral is only allowed to shrink
+void AC_PID::update_i(bool limit)
 {
-    if(!is_zero(_ki) && !is_zero(_dt)) {
-        _integrator += ((float)_input * _ki) * _dt;
-        if (_integrator < -_imax) {
-            _integrator = -_imax;
-        } else if (_integrator > _imax) {
-            _integrator = _imax;
+    if(!is_zero(_ki) && is_positive(_dt)) {
+        // Ensure that integrator can only be reduced if the output is saturated
+        if (!limit || ((is_positive(_integrator) && is_negative(_error)) || (is_negative(_integrator) && is_positive(_error)))) {
+            _integrator += ((float)_error * _ki) * _dt;
+            _integrator = constrain_float(_integrator, -_kimax, _kimax);
         }
-        _pid_info.I = _integrator;
-        return _integrator;
+    } else {
+        _integrator = 0.0f;
     }
-    return 0;
+    _pid_info.I = _integrator;
 }
 
-float AC_PID::get_d()
+float AC_PID::get_p() const
 {
-    // derivative component
-    _pid_info.D = (_kd * _derivative);
-    return _pid_info.D;
+    return _error * _kp;
 }
 
-float AC_PID::get_ff(float requested_rate)
+float AC_PID::get_i() const
 {
-    _pid_info.FF = (float)requested_rate * _ff;
-    return _pid_info.FF;
+    return _integrator;
 }
 
+float AC_PID::get_d() const
+{
+    return _kd * _derivative;
+}
 
-float AC_PID::get_pi()
+float AC_PID::get_ff()
 {
-    return get_p() + get_i();
+    _pid_info.FF = _target * _kff;
```

Comment:
It is used separately so no but I too would like to think of a more elegant way of doing this. I don't have one yet. I think it would involve providing a pointer to the PID object into the motors libary rather than just the PID and FF terms.